File: src/catter/core/js/capi/http.cc
#include <chrono>
#include <cstdint>
#include <string>
#include <unordered_map>
#include <vector>
#include <kota/http/http.h>

#include "../apitool.h"
#include "../qjs.h"

namespace qjs = catter::qjs;

namespace {

template <typename T>
using JsTask = kota::task<T, std::string>;

int64_t http_client_id_cnt = 1;
std::unordered_map<int64_t, kota::http::client> http_clients;

kota::http::client& default_http_client() {
    static kota::http::client client;
    return client;
}

kota::http::client& client_by_id(int64_t client_id) {
    auto it = http_clients.find(client_id);
    if(it == http_clients.end()) {
        throw qjs::Exception("Invalid HTTP client id: " + std::to_string(client_id));
    }
    return it->second;
}

std::vector<kota::http::header> read_headers(const qjs::Object& flat_headers) {
    auto len = flat_headers["length"].as<uint32_t>();
    if(len % 2 != 0) {
        throw qjs::Exception("HTTP headers must be a flat [name, value, ...] string array.");
    }

    std::vector<kota::http::header> headers;
    headers.reserve(len / 2);
    for(uint32_t i = 0; i < len; i += 2) {
        headers.push_back({
            .name = flat_headers[std::to_string(i)].as<std::string>(),
            .value = flat_headers[std::to_string(i + 1)].as<std::string>(),
        });
    }
    return headers;
}

qjs::Object response_to_object(JSContext* ctx, const kota::http::response& response) {
    auto result = qjs::Object::empty_one(ctx);
    result.set_property("status", response.status);
    result.set_property("ok", response.ok());
    result.set_property("url", response.url);
    result.set_property("body", response.text_copy());

    auto raw_headers = qjs::Array<std::string>::empty_one(ctx);
    for(const auto& header: response.headers) {
        raw_headers.push(header.name);
        raw_headers.push(header.value);
    }
    result.set_property("rawHeaders", qjs::Object::from(std::move(raw_headers)));

    return result;
}

JsTask<qjs::Object> send_request(JSContext* ctx,
                                 kota::http::client& client,
                                 std::string method,
                                 std::string url,
                                 qjs::Object flat_headers,
                                 std::string body,
                                 int32_t timeout_ms,
                                 int32_t max_redirects,
                                 bool danger_accept_invalid_certs,
                                 bool danger_accept_invalid_hostnames,
                                 std::string proxy_url) {
    auto request =
        client.on(kota::event_loop::current()).request(std::move(method), std::move(url));
    for(auto& header: read_headers(flat_headers)) {
        request.header(std::move(header.name), std::move(header.value));
    }

    if(!body.empty()) {
        request.body(std::move(body));
    }

    if(timeout_ms >= 0) {
        request.timeout(std::chrono::milliseconds(timeout_ms));
    }

    if(max_redirects == 0) {
        request.redirect(kota::http::redirect_policy::none());
    } else if(max_redirects > 0) {
        request.redirect(
            kota::http::redirect_policy::limited(static_cast<std::size_t>(max_redirects)));
    }

    request.danger_accept_invalid_certs(danger_accept_invalid_certs);
    request.danger_accept_invalid_hostnames(danger_accept_invalid_hostnames);

    if(!proxy_url.empty()) {
        request.proxy(std::move(proxy_url));
    }

    auto response = co_await std::move(request).send();
    if(!response) {
        co_await kota::fail(kota::http::message(response.error()));
    }

    co_return response_to_object(ctx, *response);
}

CAPI(http_client_create, ()->int64_t) {
    auto id = http_client_id_cnt++;
    http_clients.emplace(id, kota::http::client{});
    return id;
}

CAPI(http_client_close, (int64_t client_id)->void) {
    auto erased = http_clients.erase(client_id);
    if(erased == 0) {
        throw qjs::Exception("Invalid HTTP client id: " + std::to_string(client_id));
    }
}

CTX_ASYNC_CAPI(http_client_request,
               (JSContext * ctx,
                int64_t client_id,
                std::string method,
                std::string url,
                qjs::Object flat_headers,
                std::string body,
                int32_t timeout_ms,
                int32_t max_redirects,
                bool danger_accept_invalid_certs,
                bool danger_accept_invalid_hostnames,
                std::string proxy_url)
                   ->JsTask<qjs::Object>) {
    auto& client = client_by_id(client_id);
    co_return co_await send_request(ctx,
                                    client,
                                    std::move(method),
                                    std::move(url),
                                    std::move(flat_headers),
                                    std::move(body),
                                    timeout_ms,
                                    max_redirects,
                                    danger_accept_invalid_certs,
                                    danger_accept_invalid_hostnames,
                                    std::move(proxy_url));
}

CTX_ASYNC_CAPI(http_request,
               (JSContext * ctx,
                std::string method,
                std::string url,
                qjs::Object flat_headers,
                std::string body,
                int32_t timeout_ms,
                int32_t max_redirects,
                bool danger_accept_invalid_certs,
                bool danger_accept_invalid_hostnames,
                std::string proxy_url)
                   ->JsTask<qjs::Object>) {
    co_return co_await send_request(ctx,
                                    default_http_client(),
                                    std::move(method),
                                    std::move(url),
                                    std::move(flat_headers),
                                    std::move(body),
                                    timeout_ms,
                                    max_redirects,
                                    danger_accept_invalid_certs,
                                    danger_accept_invalid_hostnames,
                                    std::move(proxy_url));
}

}  // namespace

File: api/src/http.ts
import {
  http_client_close,
  http_client_create,
  http_client_request,
} from "catter-c";

export {};

export type HeaderInit = Record<string, string> | Array<[string, string]>;

export type RequestOptions = {
  method?: string;
  headers?: HeaderInit;
  body?: string;
  timeoutMs?: number;
  maxRedirects?: number;
  dangerAcceptInvalidCerts?: boolean;
  dangerAcceptInvalidHostnames?: boolean;
  proxy?: string;
};

export class Response {
  readonly status: number;
  readonly ok: boolean;
  readonly url: string;
  readonly body: string;
  readonly rawHeaders: Array<[string, string]>;
  readonly headers: Record<string, string>;

  constructor(raw: {
    status: number;
    ok: boolean;
    url: string;
    body: string;
    rawHeaders: string[];
  }) {
    this.status = raw.status;
    this.ok = raw.ok;
    this.url = raw.url;
    this.body = raw.body;
    this.rawHeaders = pairs(raw.rawHeaders);
    this.headers = normalizeHeaders(this.rawHeaders);
  }

  text(): string {
    return this.body;
  }

  json<T = unknown>(): T {
    return JSON.parse(this.body) as T;
  }

  header(name: string): string | undefined {
    return this.headers[name.toLowerCase()];
  }
}

export class Client {
  private clientId: number | undefined = http_client_create();

  close(): void {
    if (this.clientId === undefined) {
      return;
    }

    http_client_close(this.clientId);
    this.clientId = undefined;
  }

  async request(url: string, options: RequestOptions = {}): Promise<Response> {
    const clientId = this.requireClient();
    const raw = await http_client_request(
      clientId,
      options.method ?? "GET",
      url,
      flattenHeaders(options.headers),
      options.body ?? "",
      options.timeoutMs ?? -1,
      options.maxRedirects ?? -1,
      options.dangerAcceptInvalidCerts ?? false,
      options.dangerAcceptInvalidHostnames ?? false,
      options.proxy ?? "",
    );
    return new Response(raw);
  }

  get(
    url: string,
    options: Omit<RequestOptions, "method" | "body"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "GET" });
  }

  head(
    url: string,
    options: Omit<RequestOptions, "method" | "body"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "HEAD" });
  }

  del(
    url: string,
    options: Omit<RequestOptions, "method"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "DELETE" });
  }

  post(
    url: string,
    body = "",
    options: Omit<RequestOptions, "method" | "body"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "POST", body });
  }

  put(
    url: string,
    body = "",
    options: Omit<RequestOptions, "method" | "body"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "PUT", body });
  }

  patch(
    url: string,
    body = "",
    options: Omit<RequestOptions, "method" | "body"> = {},
  ): Promise<Response> {
    return this.request(url, { ...options, method: "PATCH", body });
  }

  async text(url: string, options: RequestOptions = {}): Promise<string> {
    return (await this.request(url, options)).text();
  }

  async json<T = unknown>(
    url: string,
    options: RequestOptions = {},
  ): Promise<T> {
    return (await this.request(url, options)).json<T>();
  }

  private requireClient(): number {
    if (this.clientId === undefined) {
      throw new Error("HTTP client is closed");
    }
    return this.clientId;
  }
}

let defaultClient: Client | undefined;

function getDefaultClient(): Client {
  defaultClient ??= new Client();
  return defaultClient;
}

export async function request(
  url: string,
  options: RequestOptions = {},
): Promise<Response> {
  return getDefaultClient().request(url, options);
}

export function get(
  url: string,
  options: Omit<RequestOptions, "method" | "body"> = {},
): Promise<Response> {
  return getDefaultClient().get(url, options);
}

export function head(
  url: string,
  options: Omit<RequestOptions, "method" | "body"> = {},
): Promise<Response> {
  return getDefaultClient().head(url, options);
}

export function del(
  url: string,
  options: Omit<RequestOptions, "method"> = {},
): Promise<Response> {
  return getDefaultClient().del(url, options);
}

export function post(
  url: string,
  body = "",
  options: Omit<RequestOptions, "method" | "body"> = {},
): Promise<Response> {
  return getDefaultClient().post(url, body, options);
}

export function put(
  url: string,
  body = "",
  options: Omit<RequestOptions, "method" | "body"> = {},
): Promise<Response> {
  return getDefaultClient().put(url, body, options);
}

export function patch(
  url: string,
  body = "",
  options: Omit<RequestOptions, "method" | "body"> = {},
): Promise<Response> {
  return getDefaultClient().patch(url, body, options);
}

export async function text(
  url: string,
  options: RequestOptions = {},
): Promise<string> {
  return getDefaultClient().text(url, options);
}

export async function json<T = unknown>(
  url: string,
  options: RequestOptions = {},
): Promise<T> {
  return getDefaultClient().json<T>(url, options);
}

function flattenHeaders(headers: HeaderInit | undefined): string[] {
  if (!headers) {
    return [];
  }

  if (Array.isArray(headers)) {
    return headers.flatMap(([name, value]) => [name, value]);
  }

  return Object.entries(headers).flatMap(([name, value]) => [name, value]);
}

function pairs(flatHeaders: string[]): Array<[string, string]> {
  const result: Array<[string, string]> = [];
  for (let i = 0; i + 1 < flatHeaders.length; i += 2) {
    result.push([flatHeaders[i], flatHeaders[i + 1]]);
  }
  return result;
}

function normalizeHeaders(
  headers: Array<[string, string]>,
): Record<string, string> {
  const result: Record<string, string> = {};
  for (const [name, value] of headers) {
    const key = name.toLowerCase();
    result[key] = key in result ? `${result[key]}, ${value}` : value;
  }
  return result;
}
